### DISCUSSION TITLE
Crazyflie client in a Java desktop application

### DISCUSSION THREAD

**Discussion by tallner** [bitcraze/discussions#274] [Q&A]
Hi
I have a Bitcraze Crazyflie version 2.1.
I am looking at implementing a client in Java and a small GUI in FX to control the crazyflie and to fetch some data and send to the cloud.

I downloaded the Java library from Fred02 into Eclipse with Help-->Install new software...
After that I could not figure out the next step to use it in my code, just a regular import didn't cut it.
I tried to follow the readme, but that didn't help.

Instead I tried to download the entire code from Fred02:s repo https://github.com/fredg02/se.bitcraze.crazyflie.lib, imported it into my workspace and then I got the error that
Bundle 'org.usb4java.libusb4java.win32-x86-64' cannot be resolved.

So I downloaded and added the libusb4java as an external JAR, but this did not change anything.

Is there anyone able to push me in the right direction, it feels like I am just testing a lot of different option without getting anywhere.

Thanks
Chris

**Comment by krichardsson** [comment]
@fredg02 Do you have any input on this?

**Reply by fredg02** [reply]
I'll try to answer in the coming days.

**Reply by krichardsson** [reply]
Thanks!

**Reply by tallner** [reply]
Hi
@fredg02 I am the one that emailed you the other day, but I guess it is better to keep the communication here?

I think I came a bit further, but probably not in the right way...
In Eclipse I installed the software with Help->Install New Software.
This downloads the jar to C:\Users\myUserName\.p2\pool\plugins
I added the jar to my project and it seems i can use this library now.

But I don't think this is how the workflow is supposed to be :)

Just to try one of the examples I copied the code manually from 
https://github.com/fredg02/se.bitcraze.crazyflie.lib/blob/master/se.bitcraze.crazyflie.lib.examples/src/main/java/se/bitcraze/crazyflie/lib/examples/ConsoleExample.java

I got an error on row 28:
mCrazyflie = new Crazyflie(new RadioDriver(new UsbLinkJava()));
I assume this is because I dont have the antenna connected to the USB?


However, I never managed to get 3 steps in the integration part in the readme to work.

Have a nice day!

**Reply by fredg02** [reply]
Hi Chris,

Sorry, that this is less straight forward than expected.

While trying to set this up myself, I noticed that a few things needed to be updated (e.g. the Tycho version) to get the Maven build to run. I was also able to simplify the setup for the dependencies (see the "Integration section" in the README.md). 

You _should_ be able to clone the lasted version of library repo and run a simple Maven build as described in https://github.com/fredg02/se.bitcraze.crazyflie.lib#integration. This should give you a compiled library and it's dependencies in the generated P2 repository here:
`se.bitcraze.crazyflie.lib-repository/target/repository/plugins`.

To run the [`ConsoleExample.java`](https://github.com/fredg02/se.bitcraze.crazyflie.lib/blob/master/se.bitcraze.crazyflie.lib.examples/src/main/java/se/bitcraze/crazyflie/lib/examples/ConsoleExample.java) class "stand-alone" (outside of the library repo) in Eclipse, I did the following:
1. Create a new empty Java project in Eclipse
2. Copy the [`ConsoleExample.java`](https://github.com/fredg02/se.bitcraze.crazyflie.lib/blob/master/se.bitcraze.crazyflie.lib.examples/src/main/java/se/bitcraze/crazyflie/lib/examples/ConsoleExample.java) class into the Java project
3. Create a `lib` folder in the Java project
4. Copy the following JAR files from `se.bitcraze.crazyflie.lib-repository/target/repository/plugins` to the `lib` folder:
   - com.fasterxml.jackson.core.jackson-annotations_2.13.2.v20220426-1653.jar
   - com.fasterxml.jackson.core.jackson-core_2.13.2.v20220426-1653.jar
   - com.fasterxml.jackson.core.jackson-databind_2.13.2.v20220426-1653.jar
   - org.apache.commons.lang3_3.1.0.v201403281430.jar
   - org.slf4j.api_1.7.30.v20200204-2150.jar
   - org.usb4java.bundle_1.3.0.jar
   - org.usb4java.libusb4java.linux-x86_1.3.0.jar
   - org.usb4java.libusb4java.linux-x86-64_1.3.0.jar
   - se.bitcraze.crazyflie.lib-0.0.2-SNAPSHOT.jar
   - se.bitcraze.crazyflie.lib-0.0.2-SNAPSHOT-sources.jar
   - usb4java-javax_javax.usb.api_1.3.0.jar
5. Add all JARs to the build path (Select all JARs, right click -> Build Path -> Add to Build Path)
6. You should not see any compile errors anymore
7. Try to run the example (Select `ConsoleExample.java`, right click -> Run as -> 1 Java application)
   - If you see `NoClassDefFound` errors, you will need to manually add the libraries in the Dependencies tab of the Run configuration

PLEASE NOTE (Otherwise the examples won't work!):
* Your Crazyradio needs to be connected to the USB port.
* The Crazyflie needs to be switched on.
* In the examples, the radio channel is hardcoded to channel 80. You might need to adjust it accordingly.

I have tested this on a Ubuntu 20.04 LTS machine with JDK 11 and Maven 3.8.5. If you use MacOS or Windows, you will need to use the corresponding `org.usb4java.libusb4java.*` JAR files instead.

Which IDE are you using? If you are not planning to create an application/plugin for Eclipse then you can ignore the P2 stuff. The instructions above should work similarly for other IDEs. 

Hope that helps.

**Comment by tallner** [comment]
Hi Fred,

I am running a Windows 10 laptop.
The IDE:
Eclipse IDE for Java Developers (includes Incubating components)
Version: 2022-06 (4.24.0)
Build id: 20220609-1112

I cloned the repo and ran mvn clean verify -B -f se.bitcraze.crazyflie.lib-parent/pom.xml from the command prompt.
Unfortunately the build is not generating any files/directories in se.bitcraze.crazyflie.lib-repository/target/repository/plugins

Also the tests get alot of errors.
As far as I can see they are related to the USB library somehow, like this:
-------------------------------------------------------------------------------
Test set: se.bitcraze.crazyflie.lib.bootloader.BootloaderTest
-------------------------------------------------------------------------------
Tests run: 6, Failures: 0, Errors: 4, Skipped: 0, Time elapsed: 0.672 s <<< FAILURE! - in se.bitcraze.crazyflie.lib.bootloader.BootloaderTest
testGetFlashTargets(se.bitcraze.crazyflie.lib.bootloader.BootloaderTest)  Time elapsed: 0.097 s  <<< ERROR!
java.lang.NullPointerException: Cannot invoke "javax.usb.UsbConfiguration.getUsbInterfaces()" because "activeUsbConfiguration" is null
	at se.bitcraze.crazyflie.lib.bootloader.BootloaderTest.testGetFlashTargets(BootloaderTest.java:148)

![image](https://user-images.githubusercontent.com/79505011/190900215-bf13f068-04af-4be1-9f36-227c6e020bc8.png)


Do you have any idea what the problem could be?

**Reply by fredg02** [reply]
Please skip the tests for now:
```
mvn clean verify -B -f se.bitcraze.crazyflie.lib-parent/pom.xml -DskipTests
```

**Reply by tallner** [reply]
Yes, this managed to build the project!

**Comment by fredg02** [comment]
Not sure if you have used the Crazyflie Python Client before and tested that the connection to the Crazyflie works. In case you have not, you might need to set up the Crazyflie USB Windows driver as well, see https://www.bitcraze.io/documentation/repository/crazyradio-firmware/master/building/usbwindows/.

**Reply by tallner** [reply]
No I havn't used it with my laptop yet, only with the iPhon app.
Thanks!

**Comment by tallner** [comment]
I ran the consoleExample, but I don't think it connects.

![image](https://user-images.githubusercontent.com/79505011/190916622-10c2cf30-19bd-420f-9b07-f9238b8e0f10.png)

**Reply by fredg02** [reply]
Please try `ScanExample` to see if you are using the right radio channel.

**Reply by tallner** [reply]
Radio channel was correct but the datarate was wrong.

So now it looks like the library is setup and communication is working.

Thank you! :)

![image](https://user-images.githubusercontent.com/79505011/190918903-bfcae605-36dc-494c-9204-f1695a4ffcfc.png)

**Reply by tallner** [reply]
The logging is not working though, I couldn't figure out why the logger class couldn't be loaded

**Reply by fredg02** [reply]
A logging provider is missing. You can add something like the slf4j-simple logger to make this work. I've added it as a Maven dependency to the examples pom.xml for now. See https://github.com/fredg02/se.bitcraze.crazyflie.lib/commit/2afae1835b815ba33bc01e59cabd5310a02bf514

**Reply by tallner** [reply]
This works now!
I can read & write to the Crazyflie and I should be able to start the project I am doing.
Thank you!

**Reply by fredg02** [reply]
:+1: I'm glad it works.

**Comment by tallner** [comment]
This is not a very urgent question, but can I upgrade these APIs?
Or there might not be a need for it?
Are there any documentation for the APIs?

![image](https://user-images.githubusercontent.com/79505011/190920205-986a7b9e-3a58-41f7-b30b-84ad12598727.png)

**Reply by fredg02** [reply]
Let me dig a bit. I thought I had already fixed this, but my memory might have read errors.

**Comment by tallner** [comment]
Is there any documentation for the API?
I cannot understand how it works when sending, for example, new thrust.
I thought if I send for example 25000 once, then this is the thrust until I set a new.

But It seems I need to write something like this to work and then it runs for about 3 seconds no matter what i set the first sleep to

`
this.mCrazyflie.sendPacket(new CommanderPacket(0, 0, 0, (char) 0));
        
this.mCrazyflie.sendPacket(new CommanderPacket(roll, pitch, yawrate, (char) thrust));
try {
    Thread.sleep(1000);
} catch (InterruptedException e) {
    e.printStackTrace();
}
  
this.mCrazyflie.sendPacket(new CommanderPacket(0, 0, 0, (char) 0));
// Make sure that the last packet leaves before the link is closed
// since the message queue is not flushed before closing
try {
    Thread.sleep(1000);
} catch (InterruptedException e) {
    e.printStackTrace();
}
`

**Reply by tallner** [reply]
I found out I need to continuously send roll, pitch, yaw, thrust in a separate thread.
